### DISCUSSION TITLE
Handling of TFIELDS/TFORMn

### DISCUSSION THREAD

**Discussion by keastrid** [nom-tam-fits/nom-tam-fits#483] [General]
It seems like the fpack util is producing invalid (by FITS >=3.0, FITS 2.1b doesn't specify this as a hard requirement it seems) by specifying 4 fields, but only containing 3. Has anyone seen something like this before? I've tried this on two systems (windows with a built fpack and ubuntu 20.04 with the included fpack), so I don't think its a broken install.

cfitsio/funpack appear to have no problem opening files like this. Is there a lenient mode that can be enabled in nom.tam?

Presumably, fpack is outdated and using a <=2.1b standard for compression, and is still maintaining that support for backwards compatibility with previously created files. 

nom.tam at least is producing valid files that cfitsio/funpack have no problem with.

[The file.](https://github.com/nom-tam-fits/nom-tam-fits/files/12559598/ground.fits.fz.zip)

**Comment by keastrid** [comment]
So, found where the TFIELDS=4 is from - this was an exported slice from a multiHDU compressed image. I guess I need to improved the header stripping when it gets copied over...

**Reply by attipaci** [reply]
@keastrid, is this something we need to fix in the library, or it's more of a problem with how you have been using it? If it's the former, would you like to take a shot at it?

**Reply by keastrid** [reply]
@attipaci If this is something that needs to be fixed needs some discussion, I think.

This situation poses two potential issues:
1. nom.tam's behavior diverges from cfitsio's in the handling of TFIELDS > number of TFORMn's. Given how this has been clarified in more recent versions of the FITS specification, I am unsure if nom.tam should "fix" this.
2. Lack of validation of FITS headers when writing. This is more an issue with how I've been using it. We allow the saving to disk of a FITS image from a slice (a FITS file, an image from a FITS table, or one HDU from a multiHDU image), and we try to preserve the initial header (mostly for WCS information). I might've missed an API, but it seems nom.tam doesn't handle this situation well, we strip parts of initial header before making the new one, but clearly that can be improved on our end.

**Reply by attipaci** [reply]
@keastrid Thanks for the clarification. As for part 1, I'd think it more important to follow the FITS specification than to reproduce the behavior of cfitsio. So, if we conform to the TFIELDS specification, then that is that. 

As for part 2, it probably goes back to the fact that nom.tam.fits is ultimately a very low-level library since its inception and by design. That is, it requires a lot of FITS knowledge from users to use properly. It provides the basic mechanics of reading / writing FITS files, but it also allows gross misuse if one isn't careful. I've been thinking about (and got started on) a different approach that hides all the implementation of the standard from the user, and thus making it both simpler to use, and also practically impossible to mess up. However, nom.tam.fits has never been that library, and there is a good chance it never will be either...

That said, if you find a way to make the TFIELDS handing more automatic and more robust, it would be a worthy change / addition for an upcoming release... ;-)

**Reply by keastrid** [reply]
I'll take a look and see how involved such a change would be.

Adjusting our header copy shouldn't be difficult.
